### PR TITLE
fixed travis error

### PR DESCRIPTION
**Pull request by siddhantbajaj**
Issue #1286 

Since we are using webmock for one of our external request, we cannot add factory rake task to travis.
But we should clear the database after linting the factories.

### CODE REVIEW

**Comment by Ana06** [comment]
Great! :+1: 

@siddhantbajaj can you please squash your commits into the first one? I mean just one commit _Fix Travis error_. In case that you don't know how to do it: https://ariejan.net/2011/07/05/git-squash-your-latests-commits-into-one

File: spec/support/database_cleaner.rb [hunk 1]
```diff
@@ -1,6 +1,16 @@
 RSpec.configure do |config|
-  config.before(:each) do
+  config.before(:suite) do
     DatabaseCleaner.clean_with(:truncation)
```
Comment (siddhantbajaj) [inline spec/support/database_cleaner.rb]: @Ana06 I think we can skip this as we are already doing it after linting.

Comment (Ana06) [inline spec/support/database_cleaner.rb]: but only if `ENV['OSEM_FACTORY_LINT'] != 'false'`

File: spec/support/factory_girl.rb [hunk 2]
```diff
@@ -6,6 +6,7 @@
     if ENV['OSEM_FACTORY_LINT'] != 'false'
       mock_commercial_request
       FactoryGirl.lint
+      DatabaseCleaner.clean_with(:truncation)
```
Comment (Ana06) [inline spec/support/factory_girl.rb]: Just realised, we can not truncate the database here because this will delete the seeds, or we need to load the seeds after truncating.

File: spec/support/database_cleaner.rb (hunk 1 above)
Comment (siddhantbajaj) [inline spec/support/database_cleaner.rb]: But I think that should always remain true unless someone wants to manually run the factory_girl rake task which indeed won't be possible as one of our factory build depend on internet.

File: spec/support/factory_girl.rb (hunk 2 above)
Comment (siddhantbajaj) [inline spec/support/factory_girl.rb]: So can we just just load and clean the database here only?

File: spec/support/database_cleaner.rb (hunk 1 above)
Comment (Ana06) [inline spec/support/database_cleaner.rb]: But you call want to change it to false to not lint the factories. In fact, the ideal solution would be running the linting code in a task as it is said here: https://github.com/thoughtbot/factory_girl/blob/master/GETTING_STARTED.md#linting-factories

> Running it in a before(:suite), will negatively impact the performance of your tests when running single tests.

Then we can run this task in Travis and if it success then running the tests too.
And for the `mock_commercial_request` we already have this `before_each`: https://github.com/openSUSE/osem/blob/c48c8afa51b3072af27ae76bb2ef0523bf1c50a8/spec/support/external_request.rb#L6, so as long as we keep this code it should work. And to be honest I don't know why the `mock_commercial_request` call is inside the if `ENV['OSEM_FACTORY_LINT'] != 'false'` as this is completely unrelated to the lint stuff.

Comment (siddhantbajaj) [inline spec/support/database_cleaner.rb]: > In fact, the ideal solution would be running the linting code in a task

Yes Indeed, my initial approach was the same but then I thought of doing it this way.
>And to be honest I don't know why the mock_commercial_request call is inside the if ENV['OSEM_FACTORY_LINT'] != 'false' as this is completely unrelated to the lint stuff.

I was in the same confusion but later find out that our factory is dependent on internet beacuse of this validation in which we check if the commercial is valid.
https://github.com/openSUSE/osem/blob/master/app/models/commercial.rb#L25
So we need to mock the request before linting the factories.

**Comment by siddhantbajaj** [comment]
Now the thing we need to decide is whether we should keep factory_girl.rb file and set ENV['OSEM_FACTORY_LINT']  to false by default or we can just delete this file?

File: lib/tasks/factory_girl.rake [hunk 3]
```diff
@@ -1,9 +1,11 @@
+require_relative '../../spec/support/external_request'
 namespace :factory_girl do
   desc "Verify that all FactoryGirl factories are valid"
   task lint: :environment do
     if Rails.env.test?
       begin
         DatabaseCleaner.start
+        mock_commercial_request
```
Comment (Ana06) [inline lib/tasks/factory_girl.rake]: :+1: Although I would put this in a separate commit as it is currently wrong and it is unrelated to the Travis error

File: Gemfile [hunk 4]
```diff
@@ -211,6 +211,7 @@ end
 
 group :test do
   # as test framework
+  gem 'sqlite3'
```
Comment (Ana06) [inline Gemfile]: Why do you need to add this?

File: spec/support/database_cleaner.rb (hunk 1 above)
Comment (Ana06) [inline spec/support/database_cleaner.rb]: Right

File: spec/support/factory_girl.rb [hunk 5]
```diff
@@ -6,6 +6,8 @@
     if ENV['OSEM_FACTORY_LINT'] != 'false'
       mock_commercial_request
       FactoryGirl.lint
```
Comment (Ana06) [inline spec/support/factory_girl.rb]: Now that we are doing it in a rake task I think that we can remove this file together with the `OSEM_FACTORY_LINT` as there is no point in linting the factories before each suite.

File: .travis.yml [hunk 6]
```diff
@@ -24,5 +24,6 @@ before_script:
   - mysql -u root -e 'create database osem_test;'
   - RAILS_ENV=test bundle exec rake db:migrate --trace
 script:
+  - 'bundle exec rake factory_girl:lint'
```
Comment (Ana06) [inline .travis.yml]: :1:

File: spec/support/database_cleaner.rb (hunk 1 above)
Comment (Ana06) [inline spec/support/database_cleaner.rb]: Right

File: Gemfile (hunk 4 above)
Comment (siddhantbajaj) [inline Gemfile]: Because of a strange issue i faced. When I tried to run 'bundle exec rake factory_girl:lint'  it shows me an error to include this gem. But when i tried "bundle exec rake factory_girl:lint RAILS_ENV='test'" there was no errors.
Though we check the ENV in factory_girl task but still i don't know why i got that error.

File: spec/support/factory_girl.rb (hunk 5 above)
Comment (siddhantbajaj) [inline spec/support/factory_girl.rb]: done.

File: .travis.yml (hunk 6 above)
Comment (siddhantbajaj) [inline .travis.yml]: I have changed this to "bundle exec rake factory_girl:lint RAILS_ENV='test'" so that we don't need to include that gem for test environment as well.

File: .travis.yml [hunk 7]
```diff
@@ -24,5 +24,6 @@ before_script:
   - mysql -u root -e 'create database osem_test;'
   - RAILS_ENV=test bundle exec rake db:migrate --trace
 script:
+  - "bundle exec rake factory_girl:lint RAILS_ENV='test'"
```
Comment (Ana06) [inline .travis.yml]: I would move this one line down to be executed just before the rspec tests

Comment (siddhantbajaj) [inline .travis.yml]: done. :+1:

**Comment by Ana06** [comment]
Just one more thing before we can merge it. Why did you write _temporary fix_ in the second commit message, why is it temporary?

I would just write something like: _Mock request to fix factory girl rake_

**Comment by siddhantbajaj** [comment]
> Although I would put this in a separate commit as it is **currently wrong** and it is unrelated to the Travis error

I thought you were not really satisfied with this fix so thought of naming it as temporary. :flushed:

**Comment by Ana06** [comment]
@siddhantbajaj 
> I thought you were not really satisfied with this fix so thought of naming it as temporary. :flushed:
I just meant that it doesn't work at the moment, but I am perfectly fine with the change :wink:

**Comment by Ana06** [comment]
So let's merge it. Great job :tada:
